src/backends/fvm_gpu.hpp

```diff
@@ -174,6 +185,129 @@ struct backend {
+            value_type time;    // time of crossing
+            __host__ __device__
+            friend bool operator==
+                (const threshold_crossing& lhs, const threshold_crossing& rhs)
+            {
+                return lhs.index==rhs.index && lhs.time==rhs.time;
+            }
+        };
+
+        using stack_type = gpu_stack<threshold_crossing>;
+
+        threshold_watcher() = default;
+
+        threshold_watcher(
+                const_view values,
+                const std::vector<size_type>& index,
+                const std::vector<value_type>& thresh,
+                value_type t=0):
+            values_(values),
+            index_(memory::make_const_view(index)),
+            thresholds_(memory::make_const_view(thresh)),
+            prev_values_(values),
+            is_spiking_(size()),
+            stack_(memory::make_managed_ptr<stack_type>(10*size()))
+        {
+            reset(t);
+        }
+
+        /// Remove all stored crossings that were detected in previous calls
+        /// to test()
+        void clear_crossings() {
+            stack_->clear();
+        }
+
+        /// Reset state machine for each detector.
+        /// Assume that the values in values_ have been set correctly before
+        /// calling, because the values are used to determine the initial state
+        void reset(value_type t=0) {
+            clear_crossings();
+
+            // Make host-side copies of the information needed to calculate
+            // the initial spiking state
+            auto values = memory::on_host(values_);
+            auto thresholds = memory::on_host(thresholds_);
+            auto index = memory::on_host(index_);
+
+            // calculate the initial spiking state in host memory
+            auto spiking = std::vector<size_type>(size());
+            for (auto i: util::make_span(0u, size())) {
+                spiking[i] = values[index[i]] < thresholds[i] ? 0 : 1;
+            }
+
+            // copy the initial spiking state to device memory
+            is_spiking_ = memory::on_gpu(spiking);
+
+            // reset time of last test
+            t_prev_ = t;
+        }
+
+        bool is_spiking(size_type i) const {
+            return is_spiking_[i];
+        }
+
+        const std::vector<threshold_crossing> crossings() const {
+            return std::vector<threshold_crossing>(stack_->begin(), stack_->end());
+        }
+
+        /// The time at which the last test was performed
+        value_type last_test_time() const {
+            return t_prev_;
+        }
+
+        /// Tests each target for changed threshold state
+        /// Crossing events are recorded for each threshold that
+        /// is crossed since the last call to test
+        void test(value_type t) {
+            EXPECTS(t_prev_<t);
+
+            constexpr int block_dim = 128;
+            const int grid_dim = (size()+block_dim-1)/block_dim;
+            test_thresholds<<<grid_dim, block_dim>>>(
```

Comment:
@halfflat and I have discussed what to do with the CUDA quite a bit, and we are leaning towards compiling all of the CUDA code separately (or at runtime, like you have done in past projects).
That would also save us the slow and buggy compilation of the main application with nvcc.